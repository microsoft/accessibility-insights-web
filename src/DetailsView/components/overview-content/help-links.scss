// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../../common/styles/colors.scss';
@import '../../../common/styles/fonts.scss';

.help-link {
    padding: 5px 0 5px 0;
    font-family: $font-family;
    line-height: 20px;
    font-size: 15px;
    color: $communication-primary;
}
