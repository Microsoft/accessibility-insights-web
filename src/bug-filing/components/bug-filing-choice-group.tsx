// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ChoiceGroup, IChoiceGroupOption } from 'office-ui-fabric-react/lib/ChoiceGroup';
import * as React from 'react';

import { UserConfigMessageCreator } from '../../common/message-creators/user-config-message-creator';
import { NamedSFC } from '../../common/react/named-sfc';
import { BugFilingService } from '../types/bug-filing-service';

export type BugFilingChoiceGroupProps = {
    deps: BugFilingChoiceGroupDeps;
    selectedBugFilingService: BugFilingService;
    bugFilingServices: BugFilingService[];
};

export type BugFilingChoiceGroupDeps = {
    userConfigMessageCreator: UserConfigMessageCreator;
};

export const BugFilingChoiceGroup = NamedSFC<BugFilingChoiceGroupProps>('BugFilingChoiceGroup', props => {
    const { userConfigMessageCreator } = props.deps;

    const getOptions: () => IChoiceGroupOption[] = () => {
        return props.bugFilingServices.map(service => {
            return {
                key: service.key,
                text: service.displayName,
            };
        });
    };

    const onChange = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, option?: IChoiceGroupOption) => {
        userConfigMessageCreator.setBugService(option.key);
    };

    return (
        <ChoiceGroup
            className={'bug-filing-choice-group'}
            onChange={onChange}
            options={getOptions()}
            selectedKey={props.selectedBugFilingService.key}
        />
    );
});
