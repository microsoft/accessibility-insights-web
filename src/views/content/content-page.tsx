// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import { HyperlinkDefinition } from 'common/types/hyperlink-definition';
import { flatten, forEach, toPairs } from 'lodash';
import * as React from 'react';

import { MarkupBasedComponentProps, createMarkup, Markup, MarkupDeps } from './markup';

type HyperlinkDefinitionMap = { [KEY in string]: { href: string; text: string } };
type HyperlinkComponentMap<M extends HyperlinkDefinitionMap> = {
    [KEY in keyof M]: React.FC<React.PropsWithChildren<unknown>>;
};

export type ContentPageDeps = MarkupDeps;
export type ContentPageProps = MarkupBasedComponentProps;

export type ContentPageComponent = React.FC<React.PropsWithChildren<ContentPageProps>> & {
    displayName: 'ContentPageComponent';
    pageTitle?: string;
};

export type ContentReference = string | ContentPageComponent;
type CreateProps<M extends HyperlinkDefinitionMap> = {
    Markup: Markup;
    Link: HyperlinkComponentMap<M>;
};
export function ContentCreator<M extends HyperlinkDefinitionMap>(
    linkMap?: M,
): (fn: (props: CreateProps<M>) => JSX.Element) => ContentPageComponent {
    function mapLinks(markup: Markup): HyperlinkComponentMap<M> {
        const map: Partial<HyperlinkComponentMap<M>> = {};

        forEach(linkMap, (hyperlink: HyperlinkDefinition, key: keyof M) => {
            map[key] = ({ children }) => (
                <markup.HyperLink href={hyperlink.href}>
                    {children || hyperlink.text}
                </markup.HyperLink>
            );
        });

        return map as HyperlinkComponentMap<M>;
    }

    function create(fn: (props: CreateProps<M>) => JSX.Element): ContentPageComponent {
        return NamedFC<ContentPageProps>('ContentPageComponent', props => {
            const { deps, options } = props;
            const markup = createMarkup(deps, options);
            return fn({ Markup: markup, Link: mapLinks(markup) });
        }) as ContentPageComponent;
    }

    return create;
}

export interface ContentProvider {
    getPage(path: string): ContentPageComponent;
    allPaths(): string[];
    pathTo(component: ContentPageComponent): string | null;
    contentFromReference(content?: ContentReference): ContentPageComponent;
    pathFromReference(content?: ContentReference): string | null;
}
type ContentTree = { [K in string]: ContentTree | ContentPageComponent };
export function ContentProvider(root: ContentTree): ContentProvider {
    const create = ContentCreator();

    const notFoundPage = (path?: string) => create(() => <h1>Cannot find {path}</h1>);

    function isContentPageComponent(
        leaf: any | ContentPageComponent,
    ): leaf is ContentPageComponent {
        return (
            leaf &&
            (leaf as ContentPageComponent).displayName &&
            (leaf as ContentPageComponent).displayName === 'ContentPageComponent'
        );
    }

    type TreeEntry = { path: string; leaf: ContentPageComponent };
    const flattenTree = (tree: ContentTree): TreeEntry[] => {
        const prefixEntry =
            (prefix: string) =>
            ({ path, leaf }: TreeEntry) =>
                ({
                    path: prefix + '/' + path,
                    leaf,
                }) as TreeEntry;

        const entries = toPairs(tree).map(([key, leaf]) =>
            isContentPageComponent(leaf)
                ? ({ path: key, leaf } as TreeEntry)
                : flattenTree(leaf).map(prefixEntry(key)),
        );

        return flatten(entries);
    };
    const rootEntries = flattenTree(root);

    function findPage(tree: ContentTree, [head, ...tail]: string[]): ContentPageComponent | null {
        if (!tree) {
            return null;
        }

        const branch = tree[head];
        if (isContentPageComponent(branch)) {
            if (tail.length === 0) {
                return branch;
            } else {
                return null;
            }
        } else {
            return findPage(branch, tail);
        }
    }

    function getPage(path?: string): ContentPageComponent {
        return (path && findPage(root, path.split('/'))) || notFoundPage(path);
    }

    const allPaths = () => rootEntries.map(entry => entry.path);

    const pathTo = (component: ContentPageComponent) => {
        const entry = rootEntries.find(e => e.leaf === component);
        return entry ? entry.path : null;
    };

    const contentFromReference = (reference?: ContentReference) =>
        isContentPageComponent(reference) ? reference : getPage(reference);
    const pathFromReference = (reference?: ContentReference) =>
        isContentPageComponent(reference) ? pathTo(reference) : reference ?? null;

    return { getPage, allPaths, pathTo, contentFromReference, pathFromReference };
}

export const ContentPage = {
    create: ContentCreator(),
    provider: ContentProvider,
};
