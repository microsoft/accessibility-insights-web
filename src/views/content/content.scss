// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../common/icons/icon.scss';
@import '../../common/styles/colors.scss';
@import '../../common/styles/fonts.scss';

.content-container {
    display: flex;

    .content-left,
    .content-right {
        flex-grow: 1;
        max-width: 150px;
    }
}

.content-header {
    @include text-style-title-l;

    margin-left: 10px;
    margin-right: 10px;
}

.content {
    flex-grow: 10;
    max-width: 1120px;
    font-family: $font-family;
    font-size: 14px;
    margin: 10px;
    word-break: break-word;

    table {
        font-family: $font-family;
        font-size: 14px;
    }

    h1 {
        @include text-style-title-l;
    }

    h2 {
        font-size: 21px;
        font-weight: 700;
        letter-spacing: -0.02em;
        line-height: 32px;
    }

    h3 {
        font-size: 14px;
    }

    h4 {
        font-weight: 400;
        font-size: 14px;
        margin-bottom: 4px;
    }

    h5 {
        font-weight: 700;
        font-size: 14px;
        margin-bottom: 4px;
    }

    a {
        font-weight: 400;
    }

    ol {
        -webkit-padding-start: 16px; // Chrome versions before 87 use non-standard name: https://caniuse.com/mdn-css_properties_padding-inline-start
        padding-inline-start: 16px;

        li ul li {
            list-style-type: disc;
        }
    }

    .content-hyperlinks {
        .insights-link {
            line-height: 20px;
            margin-bottom: 6px;
            display: block;
        }

        .content-inline {
            line-height: 20px;
            margin-bottom: 6px;

            .insights-link {
                display: inline;
            }
        }
    }

    .code-example {
        .code-example-code {
            font-family: $code-font-family;
            margin-block-start: 1em;
            margin-block-end: 1em;
        }

        .code-example-title {
            h4 {
                display: inline;
                margin-block-start: 0;
            }
        }
    }

    .columns {
        margin-top: 40px;
        @media screen and (width >= 800px) {
            display: flex;
        }
    }

    .column {
        padding: 10px;
        flex-grow: 1;
        flex-basis: 0;
        max-width: 520px;
    }

    .column + .column {
        @media screen and (width >= 800px) {
            margin-left: 80px;
        }
    }

    $icon-size: 16px;
    $do-color: $positive-outcome;
    $dont-color: $negative-outcome;

    .insights-code {
        font-family: $code-font-family;
    }

    .pass-fail-grid {
        display: grid;
        grid-template-areas:
            'fail-section pass-section'
            'fail-example pass-example';
        border: 1px solid $neutral-60;
        word-break: break-word;
    }

    .fail-section,
    .fail-example,
    .pass-section,
    .pass-example {
        padding: 15px;
    }

    .fail-example,
    .pass-example {
        border-top: 1px solid $neutral-60;
    }

    .fail-section,
    .fail-example {
        border-right: 1px solid $neutral-60;
    }

    .fail-section {
        grid-area: fail-section;
    }

    .fail-example {
        grid-area: fail-example;
    }

    .pass-section {
        grid-area: pass-section;
    }

    .pass-example {
        grid-area: pass-example;
    }

    .do-header,
    .pass-header,
    .dont-header,
    .fail-header {
        width: 100%;
    }

    .do-header,
    .dont-header {
        .check-container {
            margin-top: 9px;
            float: right;
        }

        h2 {
            display: inline;
        }
    }

    .pass-header,
    .fail-header {
        h3 {
            display: inline;
        }
    }

    .do-section,
    .pass-section,
    .pass-example {
        .highlight {
            color: $do-color;
            font-weight: bold;
        }
    }

    .dont-section,
    .fail-section,
    .fail-example {
        .highlight {
            color: $dont-color;
            font-weight: bold;
        }
    }

    .do-header {
        @include check-icon-styles($icon-size, 0, $do-color);

        border-bottom: 2px solid $do-color;
    }

    .dont-header {
        @include cross-icon-styles($icon-size, 0, $dont-color);

        border-bottom: 2px solid $dont-color;
    }

    .pass-header {
        @include check-icon-styles(12px, 0, $do-color);
    }

    .fail-header {
        @include cross-icon-styles(12px, 0, $dont-color);
    }

    .table {
        list-style: none;
        margin-top: 12px;
        display: table;
        border-collapse: separate;
        border-spacing: 5px;
        padding-left: 18px;
    }

    .accessibility-problems-list {
        background: $neutral-6;
        list-style-type: disc;
    }

    .landmark-table {
        border-spacing: 5px;

        @media screen and (width >= 600px) {
            padding-left: 32px;
        }

        .landmark-table-column {
            width: 24%;
        }

        .landmark-role {
            text-align: center;
            border-style: dashed;
            border-width: 3px;
        }

        td,
        th {
            padding: 5px;
            vertical-align: text-top;
        }

        th {
            background: $neutral-100;
            color: $neutral-0;
            font-weight: 400;
        }

        td {
            background: $neutral-10;
            color: $neutral-100;
        }

        .contentinfo-landmark {
            border-color: $landmark-contentinfo;
        }

        .main-landmark {
            border-color: $landmark-main;
        }

        .complementary-landmark {
            border-color: $landmark-complementary;
        }

        .banner-landmark {
            border-color: $landmark-banner;
        }

        .region-landmark {
            border-color: $landmark-region;
        }

        .navigation-landmark {
            border-color: $landmark-navigation;
        }

        .search-landmark {
            border-color: $landmark-search;
        }

        .form-landmark {
            border-color: $landmark-form;
        }
    }

    .landmarks-legend {
        color: $neutral-100;
        font-size: 14px;
        display: table-cell;
        width: 110px;
        text-align: center;
        font-weight: 400;
        padding: 4px;
        vertical-align: middle;
        border-style: dashed;
        border-width: 3px;

        &.contentinfo-landmark {
            border-color: $landmark-contentinfo;
        }

        &.main-landmark {
            border-color: $landmark-main;
        }

        &.complementary-landmark {
            border-color: $landmark-complementary;
        }

        &.banner-landmark {
            border-color: $landmark-banner;
        }

        &.region-landmark {
            border-color: $landmark-region;
        }

        &.navigation-landmark {
            border-color: $landmark-navigation;
        }

        &.search-landmark {
            border-color: $landmark-search;
        }

        &.form-landmark {
            border-color: $landmark-form;
        }
    }
}
