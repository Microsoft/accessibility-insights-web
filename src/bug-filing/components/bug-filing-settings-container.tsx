// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';
import { BugServiceProperties } from '../../common/types/store-data/user-configuration-store';
import { SettingsDeps } from '../../DetailsView/components/settings-panel/settings/settings-props';
import { BugFilingServiceProvider } from '../bug-filing-service-provider';
import { BugFilingService } from '../types/bug-filing-service';
import { BugFilingChoiceGroup, BugFilingChoiceGroupDeps } from './bug-filing-choice-group';

export interface BugFilingSettingsContainerProps {
    deps: BugFilingSettingsContainerDeps;
    selectedBugFilingService: BugFilingService;
    selectedBugFilingServiceData: BugServiceProperties;
}

export type BugFilingSettingsContainerDeps = {
    bugServiceProvider: BugFilingServiceProvider;
} & BugFilingChoiceGroupDeps &
    SettingsDeps;

export const BugFilingSettingsContainer = NamedSFC<BugFilingSettingsContainerProps>('BugFilingSettingsContainer', props => {
    const { deps, selectedBugFilingService, selectedBugFilingServiceData } = props;
    const SettingsForm = selectedBugFilingService.renderSettingsForm;
    const bugFilingServices = deps.bugServiceProvider.allVisible();

    return (
        <>
            <BugFilingChoiceGroup deps={deps} selectedBugFilingService={selectedBugFilingService} bugFilingServices={bugFilingServices} />
            <SettingsForm deps={deps} settings={selectedBugFilingServiceData} />
        </>
    );
});
