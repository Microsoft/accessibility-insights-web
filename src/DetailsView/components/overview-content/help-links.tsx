// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { HyperlinkDefinition } from 'views/content/content-page';

import { ExternalLink, ExternalLinkDeps } from '../../../common/components/external-link';
import { NamedSFC } from '../../../common/react/named-sfc';
import { helpLink } from './help-links.scss';

export type HelpLinksDeps = ExternalLinkDeps;

export interface HelpLinksProps {
    deps: HelpLinksDeps;
    linkInformation: HyperlinkDefinition[];
}

export const HelpLinks = NamedSFC('HelpLinks', (props: HelpLinksProps) => {
    const { linkInformation } = props;
    return (
        <>
            {linkInformation.map((link: HyperlinkDefinition) => (
                <div className={helpLink} key={link.href}>
                    <ExternalLink deps={props.deps} href={link.href}>
                        {link.text}
                    </ExternalLink>
                </div>
            ))}
        </>
    );
});
