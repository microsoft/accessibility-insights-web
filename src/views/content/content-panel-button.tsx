// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ActionButton } from '@fluentui/react';
import * as React from 'react';

import { ContentActionMessageCreator } from '../../common/message-creators/content-action-message-creator';
import { NamedFC } from '../../common/react/named-fc';
import { ContentProvider, ContentReference } from './content-page';

export type ContentPanelButtonDeps = {
    contentProvider: ContentProvider;
    contentActionMessageCreator: ContentActionMessageCreator;
};

export type ContentPanelButtonProps = {
    contentTitle: string;
    deps: ContentPanelButtonDeps;
    reference: ContentReference;
    iconName: string;
};

export const ContentPanelButton = NamedFC<ContentPanelButtonProps>(
    'ContentPanelButton',
    ({ contentTitle, deps, reference, children, iconName }) => {
        const { contentProvider, contentActionMessageCreator } = deps;

        if (!reference) {
            return null;
        }

        const contentPath = contentProvider.pathFromReference(reference);
        if (contentPath == null) {
            return null;
        }

        const onClick = ev =>
            contentActionMessageCreator.openContentPanel(ev, contentPath, contentTitle);

        return (
            <ActionButton
                iconProps={{ iconName }}
                onClick={onClick}
                ariaLabel={'info and examples'}
            >
                {children}
            </ActionButton>
        );
    },
);
