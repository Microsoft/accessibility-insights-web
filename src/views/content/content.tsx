// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { NamedSFC } from '../../common/react/named-sfc';
import { ContentPageDeps, ContentProvider, ContentReference } from './content-page';
import { Page } from '../page/page';

export type ContentDeps = { contentProvider: ContentProvider } & ContentPageDeps;

export type ContentProps = { deps: ContentDeps; reference: ContentReference };

export const Content = NamedSFC<ContentProps>('Content', ({ deps, reference }) => {
    const { contentProvider } = deps;
    const ContentPage = contentProvider.contentFromReference(reference);
    return (
        <Page>
            <div className="content-container">
                <div className="content-left" />
                <div className="content">
                    <ContentPage deps={deps} options={{ setPageTitle: true }} />
                </div>
                <div className="content-right" />
            </div>
        </Page>
    );
});
