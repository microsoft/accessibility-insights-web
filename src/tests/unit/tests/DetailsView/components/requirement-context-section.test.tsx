// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { render } from '@testing-library/react';
import { HelpLinks } from 'DetailsView/components/help-links';
import {
    RequirementContextSection,
    RequirementContextSectionDeps,
    RequirementContextSectionProps,
} from 'DetailsView/components/requirement-context-section';
import * as React from 'react';
import { ContentPage, ContentPageComponent } from 'views/content/content-page';
import { ContentPanelButton } from 'views/content/content-panel-button';
import {
    expectMockedComponentPropsToMatchSnapshots,
    mockReactComponents,
} from '../../../mock-helpers/mock-module-helpers';

jest.mock('views/content/content-panel-button');
jest.mock('DetailsView/components/help-links');

describe('RequirementContextSectionTest', () => {
    let props: RequirementContextSectionProps;
    beforeEach(() => {
        mockReactComponents([ContentPanelButton, HelpLinks]);
        props = {
            deps: {} as RequirementContextSectionDeps,
            infoAndExamples: { pageTitle: 'test-page-title' } as ContentPageComponent,
            whyItMatters: ContentPage.create(() => <p>WHY IT MATTERS CONTENT FOR TESTING'</p>),
        };
    });

    it('renders content from props', () => {
        const renderResult = render(<RequirementContextSection {...props} />);
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([ContentPanelButton, HelpLinks]);
    });

    it('renders with helpfulResourceLinks', () => {
        props.helpfulResourceLinks = [{ href: 'test-guidance-href', text: 'test-guidance-text' }];
        const renderResult = render(<RequirementContextSection {...props} />);
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([ContentPanelButton, HelpLinks]);
    });
});
