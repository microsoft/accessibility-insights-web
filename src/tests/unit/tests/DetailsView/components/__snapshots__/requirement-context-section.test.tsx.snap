// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequirementContextSectionTest renders content from props 1`] = `
<DocumentFragment>
  <section
    class="requirementContextContainer"
  >
    <h3
      class="contextHeading"
    >
      Why it matters
    </h3>
    <p>
      WHY IT MATTERS CONTENT FOR TESTING'
    </p>
    <h3
      class="contextHeading"
    >
      Helpful resources
    </h3>
    <div
      class="contextLinks"
    >
      <mock-contentpanelbutton
        contenttitle="Info and examples"
        deps="[object Object]"
        reference="[object Object]"
      >
        Info and examples
      </mock-contentpanelbutton>
    </div>
  </section>
</DocumentFragment>
`;

exports[`RequirementContextSectionTest renders content from props: ContentPanelButton props 1`] = `
{
  "children": "Info and examples",
  "contentTitle": "Info and examples",
  "deps": {},
  "reference": {
    "pageTitle": "test-page-title",
  },
}
`;

exports[`RequirementContextSectionTest renders content from props: HelpLinks props 1`] = `undefined`;

exports[`RequirementContextSectionTest renders with helpfulResourceLinks 1`] = `
<DocumentFragment>
  <section
    class="requirementContextContainer"
  >
    <h3
      class="contextHeading"
    >
      Why it matters
    </h3>
    <p>
      WHY IT MATTERS CONTENT FOR TESTING'
    </p>
    <h3
      class="contextHeading"
    >
      Helpful resources
    </h3>
    <div
      class="contextLinks"
    >
      <mock-contentpanelbutton
        contenttitle="Info and examples"
        deps="[object Object]"
        reference="[object Object]"
      >
        Info and examples
      </mock-contentpanelbutton>
      <mock-helplinks
        deps="[object Object]"
        linkinformation="[object Object]"
      />
    </div>
  </section>
</DocumentFragment>
`;

exports[`RequirementContextSectionTest renders with helpfulResourceLinks: ContentPanelButton props 1`] = `
{
  "children": "Info and examples",
  "contentTitle": "Info and examples",
  "deps": {},
  "reference": {
    "pageTitle": "test-page-title",
  },
}
`;

exports[`RequirementContextSectionTest renders with helpfulResourceLinks: HelpLinks props 1`] = `
{
  "deps": {},
  "linkInformation": [
    {
      "href": "test-guidance-href",
      "text": "test-guidance-text",
    },
  ],
}
`;
